pkg/logs/internal/launchers/file/launcher.go

```diff
@@ -64,7 +66,7 @@ func NewLauncher(tailingLimit int, tailerSleepDuration time.Duration, validatePo
 	return &Launcher{
 		tailingLimit:           tailingLimit,
 		fileProvider:           fileprovider.NewFileProvider(tailingLimit, wildcardStrategy),
-		tailers:                make(map[string]*tailer.Tailer),
+		tailers:                tailers.NewTailerContainer[*tailer.Tailer](),
```

Comment:
I am a bit concerned with the amount of new granular locking this brings (and new code). Would it be simpler if we had a single mutex protected slice with a snapshot of tailers that is updated once, rather than having to lock for every small operation on the tailers field. Launcher would then be able to register itself as a container with the tracker.

Reply:
> Would it be simpler if we had a single mutex protected slice with a snapshot of tailers that is updated once

Can you elaborate on this? Would this live in the launcher or in the tailer tracker?

I briefly considered having a single collection of all the tailers in `TailerTracker` however that could introduce contention when multiple launchers are trying to create/remove them (probably unlikely to have a real performance implication, but still seemed like it should be avoided). 

The idea behind `TailerContainer` is to generalize a collection of tailer that can be safely accessed both inside and outside the launcher. By moving that logic into the launcher would require us to duplicate it (a slice, and mutex) for each launcher, and that could be risky if a mistake is made in just one of the launchers.

Reply:
> So my idea was to keep a snapshot tucked away that can be accessed concurrently, while most work continues on the private copy

I see makes sense. I was generally trying to avoid having multiple copies of the same data since it's easy to introduce inconsistencies in the future especially when the state is mutated in several places (multiple launchers).

> So maybe a sync map is the best choice here. But then we probably should have all launchers use the same storage type one and drop AnyTailerContainer.

The reason for `AnyTailerContainer` existing is it allows us to use generics to specialize `TailerContainer` to the underlying tailer type (so long as it implements `Tailer`). As a result operations like `Get` and `All` within the launcher do not require a type-cast from the interface type to perform scenario specific operations. Every tailer is so different that it would be impossible to have a truly common interface between them. 

> A symptom of this is that not all launchers use a map[string]*Tailer, or map at all.

In this specific case we can either use `TailerConatiner` as is and adapt the launcher (this should work in most cases). or - we can introduce another `TailerContainer` type that implements `AnyTailerContainer` (possibly renaming `TailerConatiner` -> `MapTailerConatiner`) and plug it into `TailerTracker`.

Let me know what you think. Happy to make changes if you feel strongly about it, but I just wanted to discuss the though process behind why It's designed the way it is.